integration-tests/src/tests/client/features/flat_storage.rs

```diff
@@ -0,0 +1,132 @@
+use crate::tests::client::process_blocks::deploy_test_contract_with_protocol_version;
+use crate::tests::client::runtimes::create_nightshade_runtimes;
+use near_chain::ChainGenesis;
+use near_chain_configs::Genesis;
+use near_client::test_utils::TestEnv;
+use near_crypto::{InMemorySigner, KeyType, Signer};
+use near_primitives::test_utils::encode;
+use near_primitives::transaction::{Action, ExecutionMetadata, FunctionCallAction, Transaction};
+use near_primitives::version::ProtocolFeature;
+use near_primitives_core::config::ExtCosts;
+use near_primitives_core::hash::CryptoHash;
+use near_primitives_core::types::Gas;
+use nearcore::config::GenesisExt;
+
+/// Check that after flat storage upgrade:
+/// - value read from contract is the same;
+/// - touching trie node cost for read decreases to zero.
+#[test]
+fn test_flat_storage_upgrade() {
+    let mut genesis = Genesis::test(vec!["test0".parse().unwrap(), "test1".parse().unwrap()], 1);
+    let epoch_length = 12;
+    let old_protocol_version = ProtocolFeature::FlatStorageReads.protocol_version() - 1;
+    let new_protocol_version = old_protocol_version + 1;
+    genesis.config.epoch_length = epoch_length;
+    genesis.config.protocol_version = old_protocol_version;
+    let chain_genesis = ChainGenesis::new(&genesis);
+    let mut env = TestEnv::builder(chain_genesis)
+        .runtime_adapters(create_nightshade_runtimes(&genesis, 1))
+        .build();
+
+    // Deploy contract to state.
+    deploy_test_contract_with_protocol_version(
+        &mut env,
+        "test0".parse().unwrap(),
+        near_test_contracts::base_rs_contract(),
+        epoch_length / 3,
+        1,
+        old_protocol_version,
+    );
+
+    let signer = InMemorySigner::from_seed("test0".parse().unwrap(), KeyType::ED25519, "test0");
+    let gas = 20_000_000_000_000;
+    let tx = Transaction {
+        signer_id: "test0".parse().unwrap(),
+        receiver_id: "test0".parse().unwrap(),
+        public_key: signer.public_key(),
+        actions: vec![],
+        nonce: 0,
+        block_hash: CryptoHash::default(),
+    };
+
+    // Write key-value pair to state.
+    {
+        let write_value_action = vec![Action::FunctionCall(FunctionCallAction {
+            args: encode(&[1u64, 10u64]),
+            method_name: "write_key_value".to_string(),
+            gas,
+            deposit: 0,
+        })];
+        let tip = env.clients[0].chain.head().unwrap();
+        let signed_transaction = Transaction {
+            nonce: 10,
+            block_hash: tip.last_block_hash,
+            actions: write_value_action,
+            ..tx.clone()
+        }
+        .sign(&signer);
+        let tx_hash = signed_transaction.get_hash();
+        env.clients[0].process_tx(signed_transaction, false, false);
+        for i in 0..epoch_length / 3 {
```

Comment:
why do we need to produce `epoch_length / 3` blocks here?

Reply:
This is subtle. Added comment and linking this to https://github.com/near/nearcore/issues/8703 for future reference

I confirmed that this issue with protocol auto-upgrade is still the case. Protocol version updates to latest after I process 4 epochs. I learned from somewhere that first 2 epochs are fine.